src/library/type_context.cpp

```diff
@@ -658,7 +659,7 @@ bool type_context::should_unfold_macro(expr const & e) {
 
 optional<expr> type_context::expand_macro(expr const & e) {
     lean_assert(is_macro(e));
-    if (should_unfold_macro(e)) {
+    if (should_unfold_macro(e) && !is_name_macro(e)) {
```

Comment:
I will try to improve `whnf_head_pred`. In the new code, it will also invoke the predicate for macros.
Thus, the caller can decide whether the macro should be unfolded or not.
What do you think?